[Rebase & FF] Consolidates memory initialization mechanism for pre-DXE environment

kuqin12 opened the pull request:
## Description

This change is a preparation PR for the platform to switch to PEI less based SEC model.

Specifically, this removes the previous change where the logic will try to poke holes in the main system memory. Instead, we will move to PEI less SEC phase, because the current PEI core is incapable of handling the static carveouts gracefully.

Accordingly, the MM communication driver also needs to handle the case when the system already added the MM communication buffer to GCD.

For details on how to complete these options and their meaning refer to [CONTRIBUTING.md](https://github.com/microsoft/mu/blob/HEAD/CONTRIBUTING.md).

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

This was tested on QEMU SBSA after switching over to PEI less SEC model.

## Integration Instructions

N/A

kuqin12 committed:
[CHERRY-PICK] ArmPlatformPkg: MemoryInitPeiLib: Read PCDs into local variable

Current MemoryPeim has excessive amount of PCD get notations. This makes
the code less readable and introduced long if statements.

This change reads most of the PCD get calls into local variables and
reference the variables instead.

Signed-off-by: Kun Qin <kuqin12@gmail.com>
(cherry picked from commit 6310b11b41500cb796246d4504fcb1dc0bc3ba2e)

kuqin12 committed:
[SQUASH ON REBASE] Revert "Carve out the mm buffer region from the bulk system memory hob (#431)"

This reverts commit ea92070eb792d2817e0b069a2f2157ab5cb6cad3.

kuqin12 committed:
[CHERRY-PICK] ArmPkg: MmCommunicationDxe: Handle when the region is already reported

When the memory region for MM communication buffer is reported, the current
driver will reject further procession.

This change is added to cover such usage so that we do not need to poke holes
in the main system memory.

Signed-off-by: Kun Qin <kuqin12@gmail.com>
(cherry picked from commit 7ccc418c23be1109aba8649ca2b08ef217f1d9e6)

os-d reviewed (approved):
Did you take the third commit to edk2? i.e. updating your earlier proposal?

kuqin12 commented:
> Did you take the third commit to edk2? i.e. updating your earlier proposal?

Yeah, I updated the PR in EDK2 to reflect this change. I think I will leave this PR open for a bit in case the EDK2 PR has other feedback that does not align on this approach.

kuqin12 committed:
ArmPlatformPkg: MemoryInitPei: Properly check resource hob v2

The previous round of change that updates the logic to handle v2 hobs
missed the processing of incoming v2 hobs that may cover the main system
memory. This change is added to cover such usage.

Signed-off-by: Kun Qin <kuqin12@gmail.com>
